ddtrace/contrib/asgi/middleware.py

```diff
@@ -0,0 +1,118 @@
+import sys
+
+from ddtrace import tracer as global_tracer
+from ddtrace.constants import ANALYTICS_SAMPLE_RATE_KEY
+from ddtrace.ext import SpanTypes, http
+from ddtrace.http import store_request_headers, store_response_headers
+from ddtrace.propagation.http import HTTPPropagator
+from ddtrace.settings import config
+
+from ...compat import reraise
+from ...internal.logger import get_logger
+from .utils import guarantee_single_callable
+
+log = get_logger(__name__)
+
+config._add("asgi", dict(service_name=config._get_service(default="asgi"), distributed_tracing=True,))
+
+ASGI_VERSION = "asgi.version"
+ASGI_SPEC_VERSION = "asgi.spec_version"
+
+
+def _extract_tags_from_scope(scope):
+    tags = {}
+
+    http_method = scope.get("method")
+    if http_method:
+        tags[http.METHOD] = http_method
+
+    if config.asgi.trace_query_string:
+        query_string = scope.get("query_string")
+        if len(query_string) > 0:
+            if isinstance(query_string, bytes):
+                query_string = query_string.decode()
+            tags[http.QUERY_STRING] = query_string
+    else:
+        query_string = None
+
+    server = scope.get("server")
+    if server and len(server) == 2:
+        port = server[1]
+        server_host = server[0] + (":" + str(port) if port != 80 else "")
+        full_path = scope.get("root_path", "") + scope.get("path", "")
+        http_url = scope.get("scheme", "http") + "://" + server_host + full_path
+        tags[http.URL] = http_url
+
+    http_version = scope.get("http_version")
+    if http_version:
+        tags[http.VERSION] = http_version
+
+    scope_asgi = scope.get("asgi")
+
+    if scope_asgi and "version" in scope_asgi:
+        tags[ASGI_VERSION] = scope_asgi["version"]
+
+    if scope_asgi and "spec_version" in scope_asgi:
+        tags[ASGI_SPEC_VERSION] = scope_asgi["spec_version"]
+
+    return tags
+
+
+class TraceMiddleware:
+    """
+    ASGI application middleware that traces the requests.
+
+    Args:
+        app: The ASGI application.
+        tracer: Custom tracer. Defaults to the global tracer.
+    """
+
+    def __init__(self, app, tracer=global_tracer):
+        self.app = guarantee_single_callable(app)
+        self.tracer = tracer
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] != "http":
+            return await self.app(scope, receive, send)
+
+        headers = {k: v for (k, v) in scope["headers"]}
```

Comment:
A note: the ASGI spec mentions that `headers` keys/values are bytes, so just wanted to make sure this is supported by the rest of processing here (eg `propagator.extract(headers)`).